bilibili: enable hardware video decode & improve wayland experience.

dingdang66686 opened the pull request:
To enable vaapi video acceleration.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kashw2 commented on pkgs/by-name/bi/bilibili/package.nix:
```diff
@@ -32,6 +33,10 @@ stdenv.mkDerivation {
     dpkg
   ];
 
+  buildInputs = [
+    libva
+  ];
```
I haven't checked the build process for Bilibili but unless there is a check for this library inside of it's build process at build time, all this will do is provide the library to the builder but not the application itself. For that you'd have to add it to it's PATH

dingdang66686 replied:
buidInputs is for runtime dependency, as described in the [wiki](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-buildInputs).

Bot-wxt1221 replied:
@dingdang66686 You should set LD_LIBRARY_PATH when wrapping.

Bot-wxt1221 commented:
This is completely stupid.

https://github.com/msojocs/bilibili-linux/issues/107

dingdang66686 commented:
Use this commandLineArg to test out: `--enable-features=AcceleratedVideoDecodeLinuxGL`, according to [ArchWiki](https://wiki.archlinux.org/title/Chromium#Hardware_video_acceleration), use wayland to get better performance: `--ozone-platform=wayland`.
<img width="1801" height="1016" alt="image" src="https://github.com/user-attachments/assets/e67d3641-8442-46f4-8a68-f089f3b4cd57" />

dingdang66686 commented:
Run electron chrome://gpu with or without prefix and flags to check if video acceleration working.

dingdang66686 commented:
Without libva no video acceleration information is shown.

Bot-wxt1221 commented:
buildInputs is not useful remove it.

Could you add your flags to makeWrapper

Bot-wxt1221 reviewed (changes requested):
up

SuperSandro2000 commented on pkgs/by-name/bi/bilibili/package.nix:
```diff
@@ -32,6 +33,10 @@ stdenv.mkDerivation {
     dpkg
   ];
 
+  buildInputs = [
+    libva
+  ];
+
```
```suggestion
```

GaetanLepage commented:
Please squash (at least) the two commits related to `libva`.

SuperSandro2000 reviewed (dismissed):
Please use the standard pattern as seen eg here https://github.com/NixOS/nixpkgs/blob/a421ac6595024edcfbb1ef950a3712b89161c359/pkgs/applications/audio/plexamp/default.nix#L33

dingdang66686 commented:
> There are also some other issues, can they be fixed here as well? For example, the command
> 
> ```shell
> sed -i "s|Exec=.*|Exec=$out/bin/bilibili|" $out/share/applications/*.desktop
> ```
> 
> removes the trailing `%u` parameter. It might be better to use `substituteInPlace` instead.
> 
> In the upstream startup script, there are two parameters: `ELECTRON_FORCE_IS_PACKAGED=true` and `--no-sandbox`. There’s also a `cursor-tool` executable included, which suggests that `autoPatchelfHook` should probably be added.

@qzylinra If I don't get it wrong, `autoPatchelfHook` only works for elf format. Why is it needed for an electron application?

qzylinra commented:
need format file

dingdang66686 committed:
bilibili: Improve user experience.

- Enable hardware video decode.
- Improve wayland experience.
- Set default environment variables for Electron.

uninsane commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 432832`
Commit: `e8ae7508cae97e01dc22a813c19dc04c891cbf00`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bilibili</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>
